### PR TITLE
Bull Leverage Rebalance

### PR DESCRIPTION
**Pull request by daryakaviani**
**Should be merged after #713**

Leverage Rebalance for Bull Strategy

### CODE REVIEW

File: packages/bull-vault/src/AuctionBull.sol [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        transferOwnership(_auctionOwner);
+    }
+
+    /**
+     * @dev changes the leverage component composition by buying or selling eth
+     */
+    function leverageRebalance(bool _isSellingEth, uint256 _amountIn, uint256 _minAmountOut, uint24 _poolFee) external {
+        if (_isSellingEth) {
+            // Withdraw ETH from collateral
+            IBullStrategy(bullStrategy).withdrawWethFromEuler(_amountIn);
```
Comment (alpinechicken) [inline packages/bull-vault/src/AuctionBull.sol:57]: If we’re doing this for usdc then maybe just calculate:

usdcToTrade = debt - (crab price)*(crab amount)

Then exactIn or exactOut to get that?

**Review by alpinechicken — CHANGES_REQUESTED** [review]
We're targeting debt = crab value in $ so maybe we can rely on the contract to calculate 

usdcToTrade = debt - (crab value in $)  

Then do exactIn if we're selling usdc or exactOut if we're buying usdc to repay loan.

File: packages/bull-vault/src/AuctionBull.sol [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        auctionManager = _auctionManager;
+
+        transferOwnership(_auctionOwner);
+    }
+
+    /**
+     * @dev changes the leverage component composition by buying or selling eth
+     */
+    function leverageRebalance(bool _isSellingEth, uint256 _amountIn, uint256 _minAmountOut, uint24 _poolFee) external {
+        if (_isSellingEth) {
```
[the remark this answers is not in the document]
Comment (daryakaviani) [reply packages/bull-vault/src/AuctionBull.sol]: Fixed

**Review by alpinechicken — COMMENTED** [review]
Why do we have params structure for flashWithdraw but all separate input params for flashDeposit?

**Comment by haythem96** [comment]
> Why do we have params structure for flashWithdraw but all separate input params for flashDeposit?

Good question, we have a task to refactor all to use a struct. I used a struct in flashWithdraw because of stack too deep error

File: packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol [hunk 3]
```diff
@@ -87,17 +90,55 @@ contract BullStrategyTestFork is Test {
...
+        assertEq(bullCrabBalanceAfter.sub(crabToDepositInitially), bullCrabBalanceBefore);
+        assertEq(bullStrategy.balanceOf(user1), crabToDepositInitially);
+        assertEq(IEulerDToken(dToken).balanceOf(address(bullStrategy)), usdcToBorrow);
+        assertTrue(wethToLend.sub(IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy))) <= 1);
+        assertEq(IERC20(usdc).balanceOf(user1), usdcToBorrow);
+
+        bullCrabBalanceBefore = bullStrategy.getCrabBalance();
+        uint256 userUsdcBalanceBefore = IERC20(usdc).balanceOf(user1);
+        uint256 userBullBalanceBefore = bullStrategy.balanceOf(user1);
+        uint256 crabToDepositSecond = 7e18;
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol:127]: going to fuzz this one?

File: packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol [hunk 4]
```diff
@@ -87,17 +90,55 @@ contract BullStrategyTestFork is Test {
...
+        (uint256 wethToLend, uint256 usdcToBorrow) = _deposit(crabToDepositInitially);
+        vm.stopPrank();
+
+        uint256 bullCrabBalanceAfter = bullStrategy.getCrabBalance();
+
+        assertEq(bullCrabBalanceAfter.sub(crabToDepositInitially), bullCrabBalanceBefore);
+        assertEq(bullStrategy.balanceOf(user1), crabToDepositInitially);
+        assertEq(IEulerDToken(dToken).balanceOf(address(bullStrategy)), usdcToBorrow);
+        assertTrue(wethToLend.sub(IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy))) <= 1);
+        assertEq(IERC20(usdc).balanceOf(user1), usdcToBorrow);
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol:122]: Check weth balance change? (maybe more so for withdraw)

File: packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol [hunk 5]
```diff
@@ -87,17 +90,55 @@ contract BullStrategyTestFork is Test {
...
+        vm.startPrank(user1);
+        (uint256 wethToLend, uint256 usdcToBorrow) = _deposit(crabToDepositInitially);
+        vm.stopPrank();
+
+        uint256 bullCrabBalanceAfter = bullStrategy.getCrabBalance();
+
+        assertEq(bullCrabBalanceAfter.sub(crabToDepositInitially), bullCrabBalanceBefore);
+        assertEq(bullStrategy.balanceOf(user1), crabToDepositInitially);
+        assertEq(IEulerDToken(dToken).balanceOf(address(bullStrategy)), usdcToBorrow);
+        assertTrue(wethToLend.sub(IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy))) <= 1);
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol:121]: Why not test ==?

File: packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol [hunk 6]
```diff
@@ -87,17 +90,55 @@ contract BullStrategyTestFork is Test {
...
+        vm.startPrank(user1);
+        (uint256 wethToLendSecond, uint256 usdcToBorrowSecond) = _deposit(crabToDepositSecond);
+        vm.stopPrank();
+
+        bullCrabBalanceAfter = bullStrategy.getCrabBalance();
+
+        assertEq(bullCrabBalanceAfter.sub(crabToDepositSecond), bullCrabBalanceBefore);
+        assertEq(bullStrategy.balanceOf(user1).sub(userBullBalanceBefore), bullToMint);
+        assertEq(IEulerDToken(dToken).balanceOf(address(bullStrategy)).sub(usdcToBorrow), usdcToBorrowSecond);
+        assertTrue(wethToLendSecond.sub(IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy)).sub(wethToLend)) <= 1);
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol:138]: Why not an == test? If it's a rounding thing then maybe abs(eToken balance - wethToLend - wethToLendSecond) <= 1?

File: packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol [hunk 7]
```diff
@@ -87,17 +90,55 @@ contract BullStrategyTestFork is Test {
...
+        assertEq(IERC20(usdc).balanceOf(user1).sub(usdcToBorrowSecond), userUsdcBalanceBefore);
+    }
+
     function testWithdraw() public {
         uint256 crabToDeposit = 15e18;
         uint256 bullToMint = _calcBullToMint(crabToDeposit);
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol:144]: bullToBurn?

File: packages/bull-vault/src/AuctionBull.sol [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ * @author opyn team
+ */
+contract AuctionBull is UniFlash, Ownable {
+    /// @dev the highest delta we can have without rebalancing
+    uint256 public deltaUpper;
+    /// @dev the lowest delta we can have without rebalancing
+    uint256 public deltaLower;
+    /// @dev the highest CR we can have before rebalancing
+    uint256 public crUpper;
+    /// @dev the lowest CR we can have in leverage component before rebalancing
```
[the remark this answers is not in the document]
Comment (daryakaviani) [reply packages/bull-vault/src/AuctionBull.sol:23]: These are deleted in https://github.com/opynfinance/squeeth-monorepo/pull/700 and replaced with guardrails.

File: packages/bull-vault/src/AuctionBull.sol (hunk 1 above)
Comment (daryakaviani) [reply packages/bull-vault/src/AuctionBull.sol]: Decided in Slack that we just take in `usdcAmount`.

File: packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol (hunk 3 above)
Comment (daryakaviani) [reply packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol:127]: This is @haythem96's deposit PR, not relevant to Leverage Rebalance.

File: packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol (hunk 5 above)
Comment (daryakaviani) [reply packages/bull-vault/test/integration-test/BullStrategyTestFork.t.sol:121]: Euler gives us +- 1 wei difference.

File: packages/bull-vault/src/AuctionBull.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (daryakaviani) [reply packages/bull-vault/src/AuctionBull.sol:23]: These values are removed in a future PR

File: packages/bull-vault/src/AuctionBull.sol [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+contract AuctionBull is UniFlash, Ownable {
+    /// @dev highest delta the auction manager can rebalance to
+    uint256 internal constant DELTA_UPPER = 1.1e18;
+    /// @dev lowest delta the auction manager can rebalance to
+    uint256 internal constant DELTA_LOWER = 0.9e18;
+    /// @dev highest CR the auction manager can rebalance to
+    uint256 internal constant CR_UPPER = 3e18;
+    /// @dev lowest CR the auction manager can rebalance to
+    uint256 internal constant CR_LOWER = 1.5e18;
+
```
Comment (haythem96) [inline packages/bull-vault/src/AuctionBull.sol:25]: We decided those as constants or we need setters?

File: packages/bull-vault/src/AuctionBull.sol [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        usdc = IController(IBullStrategy(_bull).powerTokenController()).quoteCurrency();
+        auctionManager = _auctionManager;
+
+        transferOwnership(_auctionOwner);
+    }
+
+    /**
+     * @dev changes the leverage component composition by buying or selling eth
+     */
+    function leverageRebalance(bool _isBuyingUsdc, uint256 _usdcAmount, uint256 _ethThresholdAmount, uint24 _poolFee) external {
```
Comment (haythem96) [inline packages/bull-vault/src/AuctionBull.sol:52]: Add natspec comments

File: packages/bull-vault/src/AuctionBull.sol [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        require(msg.sender == auctionManager);
```
Comment (haythem96) [inline packages/bull-vault/src/AuctionBull.sol:53]: Add revert error, and also let's use short message like `AB0`, `AB1`... etc

File: packages/bull-vault/src/AuctionBull.sol [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        
+        if (_isBuyingUsdc) {
+            // swap ETH to USDC
+            _exactOutFlashSwap(
+                weth,
+                usdc,
+                _poolFee,
+                _usdcAmount,
+                _ethThresholdAmount,
+                uint8(FLASH_SOURCE.BUYING_USDC),
```
Comment (haythem96) [inline packages/bull-vault/src/AuctionBull.sol:63]: I think flash source should always indicate from where this callback initially was executed, I would prefer a name like `FLASH_SOURCE.LELVERAGE_REBALANCE_BUYING_USDC`

File: packages/bull-vault/src/AuctionBull.sol [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            // Borrow more USDC debt
+            IBullStrategy(bullStrategy).borrowUsdcFromEuler(_usdcAmount);
+            // swap USDC to ETH 
+            _exactInFlashSwap(
+                usdc,
+                weth,
+                _poolFee,
+                _usdcAmount,
+                _ethThresholdAmount,
+                uint8(FLASH_SOURCE.SELLING_USDC),
```
Comment (haythem96) [inline packages/bull-vault/src/AuctionBull.sol:80]: same comment here for flash_source naming

File: packages/bull-vault/src/BullStrategy.sol [hunk 14]
```diff
@@ -159,4 +170,48 @@ contract BullStrategy is ERC20, LeverageBull {
...
+        );
+        uint256 crabUsdPrice = (
+            ethInCrab.wmul(ethUsdPrice).sub(
+                squeethInCrab.wmul(squeethEthPrice).wmul(ethUsdPrice)
+            )
+        ).wdiv(IERC20(crab).totalSupply());
+
+        uint256 usdcDebt; 
+        uint256 ethInCollateral;
+        uint256 expectedEth = _usdcAmount.wdiv(ethUsdPrice).mul(1e12);
```
Comment (haythem96) [inline packages/bull-vault/src/BullStrategy.sol:199]: Should we use TWAP here for `ethUsdPrice` ?

File: packages/bull-vault/src/BullStrategy.sol [hunk 15]
```diff
@@ -159,4 +170,48 @@ contract BullStrategy is ERC20, LeverageBull {
 
         return (strategyVault.collateralAmount, strategyVault.shortAmount);
     }
+
+    // TODO: Take in crab params when we add full rebalance
+    function _calcDeltaAndCR(bool _isBuyingUsdc, uint256 _usdcAmount) internal view returns (uint256, uint256) {
```
Comment (haythem96) [inline packages/bull-vault/src/BullStrategy.sol:175]: I don't think this function should be here, we need to find a way to refactor it and move some of it to LeverageRebalance()

Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:175]: I think we should move this to the AuctionBull

File: packages/bull-vault/src/LeverageBull.sol [hunk 16]
```diff
@@ -41,6 +42,8 @@ contract LeverageBull {
     address internal immutable eToken;
     /// @dev euler dToken that represent the borrowed asset
     address internal immutable dToken;
+    /// @dev auction contract address
+    address public auction;
```
Comment (haythem96) [inline packages/bull-vault/src/LeverageBull.sol:46]: We need to add setter for this

File: packages/bull-vault/src/LeverageBull.sol [hunk 17]
```diff
@@ -50,7 +53,7 @@ contract LeverageBull {
      * @param _eulerMarkets euler markets module address
      * @param _powerTokenController wPowerPerp controller address
      */
-    constructor(address _euler, address _eulerMarkets, address _powerTokenController) {
+    constructor(address _owner, address _euler, address _eulerMarkets, address _powerTokenController) Ownable() {
```
Comment (haythem96) [inline packages/bull-vault/src/LeverageBull.sol:56]: Add transferOwnership() to `_owner`

**Review by haythem96 — CHANGES_REQUESTED** [review]
Left some comments

File: packages/bull-vault/src/AuctionBull.sol [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+        }
+
+        _isValidLeverageRebalance();
+    }
+
+    function _uniFlashSwap(UniFlashswapCallbackData memory _uniFlashSwapData) internal override {
+        if (FLASH_SOURCE(_uniFlashSwapData.callSource) == FLASH_SOURCE.SELLING_USDC) {
+            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(IERC20(weth).balanceOf(address(this)), _uniFlashSwapData.amountToPay);
+
+            IERC20(_uniFlashSwapData.tokenIn).transfer(
```
Comment (aleone) [inline packages/bull-vault/src/AuctionBull.sol:132]: any reason to not use usdc, isnt it always usdc?

File: packages/bull-vault/src/AuctionBull.sol [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+                _uniFlashSwapData.pool, _uniFlashSwapData.amountToPay
+            );
+
+            IERC20(usdc).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
```
Comment (aleone) [inline packages/bull-vault/src/AuctionBull.sol:136]: how is this passing tests?

this is a duplicate transfer

File: packages/bull-vault/src/LeverageBull.sol [hunk 20]
```diff
@@ -17,6 +17,7 @@ import { UniOracle } from "./UniOracle.sol";
 /**
  * Error codes
  * LB0: ETH sent is greater than ETH to deposit in Euler
+ * LB1:
```
Comment (aleone) [inline packages/bull-vault/src/LeverageBull.sol:20]: need error string here

**Review by aleone — COMMENTED** [review]
initial review - haven't reviewed tests yet

File: packages/bull-vault/src/AuctionBull.sol [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+        transferOwnership(_auctionOwner);
+    }
+
+    /**
+     * @dev changes the leverage component composition by buying or selling eth
+     */
+    function leverageRebalance(
+        bool _isBuyingUsdc,
+        uint256 _usdcAmount,
+        uint256 _wethThresholdAmount,
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/AuctionBull.sol]: changed

File: packages/bull-vault/test/fuzz-test/FlashBullFuzzTest.t.sol [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+        // this is a crab whale, get some crab token from
+        vm.prank(0x06CECFbac34101aE41C88EbC2450f8602b3d164b);
+        IERC20(crabV2).transfer(user1, 100e18);
+        // some WETH and USDC rich address
+        vm.prank(0x57757E3D981446D585Af0D9Ae4d7DF6D64647806);
+        IERC20(weth).transfer(user1, 10100e18);
+    }
+
+    function testFuzzingFlashDeposit(uint256 _ethToCrab) public {
+        _ethToCrab = bound(_ethToCrab, 1e16, 600e18);
```
Comment (alpinechicken) [inline packages/bull-vault/test/fuzz-test/FlashBullFuzzTest.t.sol:123]: why 600 upper bound?

**Review by alpinechicken — APPROVED** [review]
lgtm - couple of comments on user eth/usdc asserts on flash deposit and withdraw but I guess these are going in a different pr now. One question is why _leverageDeposit takes _ethAmount as a param in LeverageBull? Seems redundant. Just as extra sanity check?

**Review by aleone — COMMENTED** [review]
small comments, but solidity lgtm.    need to spend a bit more time on the integration-test.

File: packages/bull-vault/test/fuzz-test/FlashBullFuzzTest.t.sol (hunk 22 above)
Comment (haythem96) [reply packages/bull-vault/test/fuzz-test/FlashBullFuzzTest.t.sol:123]: I don't know, will look at it later, this is outside of this PR scope

File: packages/bull-vault/test/unit-test/mock/Weth9Mock.t.sol [hunk 23]
```diff
@@ -15,6 +15,10 @@ contract WETH9Mock is ERC20, Test {
 
     constructor() ERC20("WETH9", "wETH9") { }
 
+    function testToAvoidCoverage() public pure {
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/test/unit-test/mock/Weth9Mock.t.sol:18]: I added thos mocks file when I was planning to initially use local deployment/unit-test, then expectRevert() worked with fork mode, so kept those files.
Now those files appear in coverage report, found this hacky way for coverage not to show those contracts.

**Review by alpinechicken — APPROVED** [review]
lgtm

**Comment by haythem96** [comment]
> _leverageDeposit

_leverageDeposit() take `_ethAmount` to make sure user don't deposit more ETH and then it is stuck, instead of going the path of checking for excess and refunding, I guess requiring sending strictly an amount of ETH equal to `wethToLend` is better

**Pull request merged**